sklearn/preprocessing/discretization.py

```diff
@@ -140,28 +150,61 @@ def fit(self, X, y=None):
         valid_encode = ('onehot', 'onehot-dense', 'ordinal')
         if self.encode not in valid_encode:
             raise ValueError("Valid options for 'encode' are {}. "
-                             "Got 'encode = {}' instead."
+                             "Got encode={!r} instead."
                              .format(valid_encode, self.encode))
+        valid_strategy = ('uniform', 'quantile', 'kmeans')
+        if self.strategy not in valid_strategy:
+            raise ValueError("Valid options for 'strategy' are {}. "
+                             "Got strategy={!r} instead."
+                             .format(valid_strategy, self.strategy))
 
         n_features = X.shape[1]
         ignored = self._validate_ignored_features(n_features)
         self.transformed_features_ = np.delete(np.arange(n_features), ignored)
 
-        offset = np.min(X, axis=0)
-        offset[ignored] = 0
-        self.offset_ = offset
         self.n_bins_ = self._validate_n_bins(n_features, ignored)
 
-        ptp = np.ptp(X, axis=0)
-        same_min_max = np.where(ptp == 0)[0]
-        if len(same_min_max) > 0:
-            warnings.warn("Features {} are constant and will be replaced with "
-                          "0.".format(", ".join(str(i) for i in same_min_max)))
+        bin_edges = np.zeros(n_features, dtype=object)
+        for jj in range(n_features):
+            if jj in ignored:
+                bin_edges[jj] = np.array([])
+                continue
+            col = X[:, jj][:, None]
+
+            if self.strategy == 'uniform':
+                edges = np.linspace(col.min(), col.max(), self.n_bins_[jj] + 1)
+
+            elif self.strategy == 'quantile':
+                quantiles = np.linspace(0, 100, self.n_bins_[jj] + 1)
+                if np_version < (1, 9):
+                    quantiles = list(quantiles)
+                edges = np.asarray(np.percentile(col, quantiles))
+
+            elif self.strategy == 'kmeans':
+                from ..cluster import KMeans
```

Comment:
Any specific reason to import here?

Reply:
Importing loop problem, but it seems not to happen anymore so I moved it to the top.